### PR TITLE
Adapt existing entity extractors and intent classifiers

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
- Modify all existing entity extractors and intent classifiers to make use of the new featurizers (e.g. sequence of features instead of just one feature vector)
- Did not modify `EmbeddingIntentClassifier` yet, will happen in a separate PR
- Use `text_sparse_features` instead of `ner_features` in `CRFEntityExtractor` as additional features. As soon as `word_emedding` is added to the list of features for the `CRFEntityExtractor` and a corresponding dense featurizer was used, the word embeddings are used as additional features for the entity extraction.

part of https://github.com/RasaHQ/research/issues/54

merged into `combined-entity-intent-model`

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 1]
```diff
@@ -22,11 +30,19 @@
     from spacy.tokens import Doc
 
 
+class CRFToken(NamedTuple):
+    text: Text
+    tag: Text
+    entity: Text
+    pattern: Dict[Text, Any]
+    word_embedding: np.ndarray
```
Comment (Ghostvv) [inline rasa/nlu/extractors/crf_entity_extractor.py]: should we use `namedtuple` directly?
`CRFToken = namedtuple("CRFToken", ("text", "tag", "entity", "pattern", "word_embedding"))`

Comment (Ghostvv) [reply rasa/nlu/extractors/crf_entity_extractor.py]: I think we discussed this at some point with @tmbo and decided on later

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 2]
```diff
@@ -582,8 +570,10 @@ def _bilou_tags_from_offsets(tokens, entities, missing="O"):
 
     @staticmethod
     def __pattern_of_token(message, i):
-        if message.get("tokens") is not None:
-            return message.get("tokens")[i].get("pattern", {})
+        if message.get(MESSAGE_TOKENS_NAMES[MESSAGE_TEXT_ATTRIBUTE]) is not None:
+            return message.get(MESSAGE_TOKENS_NAMES[MESSAGE_TEXT_ATTRIBUTE])[i].get(
```
Comment (Ghostvv) [inline rasa/nlu/extractors/crf_entity_extractor.py]: 🙈

File: CHANGELOG.rst [hunk 3]
```diff
@@ -18,10 +18,13 @@ Changed
 -------
 - Divided featurizers in ``rasa.nlu`` into sparse and dense featurizers
 - All featurizers in ``rasa.nlu`` return a sequence
+- Renamed the feature name ``ner_features`` to ``word_embedding`` in ``CRFEntityExtractor``. The ``word_embedding`` are
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply CHANGELOG.rst]: not sure, what is more intuitive for users though

File: rasa/nlu/featurizers/featurzier.py [hunk 4]
```diff
@@ -10,6 +10,21 @@
 )
 
 
+def sequence_to_sentence_embedding(
+    features: Union[List[List[float]], scipy.sparse.spmatrix], method: Text = "mean"
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/featurzier.py]: why don't we keep dense features as `np.ndarray`?

File: rasa/nlu/featurizers/featurzier.py [hunk 5]
```diff
@@ -10,6 +10,21 @@
...
+) -> Optional[np.ndarray]:
+    if features is None:
+        return None
+
+    if isinstance(features, scipy.sparse.spmatrix):
+        features = features.toarray()
+
+    if method == "mean" or method == "avg":
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/featurzier.py]: I don't think we need to support this ambiguity, let's keep it `mean` and raise exception if the string is not supported

File: rasa/nlu/featurizers/featurzier.py [hunk 6]
```diff
@@ -10,6 +10,21 @@
 )
 
 
+def sequence_to_sentence_embedding(
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/featurzier.py]: not sure about the name `embedding`

**Comment by Ghostvv** [comment]
I think `case_sensitive` was added to whitespace tokenizer for `crf` entity extractor

File: rasa/nlu/featurizers/featurzier.py (hunk 4 above)
Comment (tabergma) [reply rasa/nlu/featurizers/featurzier.py]: we do, will change the type

File: rasa/nlu/classifiers/sklearn_intent_classifier.py [hunk 7]
```diff
@@ -79,6 +81,22 @@ def transform_labels_num2str(self, y: np.ndarray) -> np.ndarray:
 
         return self.le.inverse_transform(y)
 
+    def combine_features(self, message: Message) -> np.ndarray:
+        features_1 = sequence_to_sentence_embedding(
+            message.get(MESSAGE_VECTOR_DENSE_FEATURE_NAMES[MESSAGE_TEXT_ATTRIBUTE])
+        )
+        features_2 = sequence_to_sentence_embedding(
+            message.get(MESSAGE_VECTOR_SPARSE_FEATURE_NAMES[MESSAGE_TEXT_ATTRIBUTE])
+        )
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py]: true, so should we only use sparse features here?

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py]: only dense

[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/classifiers/sklearn_intent_classifier.py]: yes, always get confused 🙈

File: CHANGELOG.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply CHANGELOG.rst]: yeah, the names are not perfect. also dense features do not have to be equal to pre-trained word embeddings. we just assume they are.

File: rasa/nlu/featurizers/featurzier.py (hunk 6 above)
Comment (tabergma) [reply rasa/nlu/featurizers/featurzier.py]: changed it to `sequence_to_sentence_features`, better?

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 1 above)
Comment (tabergma) [reply rasa/nlu/extractors/crf_entity_extractor.py]: The benefit of `class CRFToken(NamedTuple)` is that you can define types.

File: CHANGELOG.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply CHANGELOG.rst]: Will rename it to `text_dense_features` (same name as the corresponding featurizer is using). We say in the documentation that those can, for example, be pre-trained word embeddings.

File: rasa/nlu/featurizers/featurzier.py (hunk 6 above)
Comment (Ghostvv) [reply rasa/nlu/featurizers/featurzier.py]: do we plan to use this method further? Currently it converts sparse to dense, but we'd like to make this conversion latter during batching

Comment (tabergma) [reply rasa/nlu/featurizers/featurzier.py]: the plan is to keep this method for the `SklearnIntentClassifier`, for example. So, for extractors/classifiers that cannot handle a sequence. We can convert it back to sparse if needed.

File: rasa/nlu/featurizers/featurzier.py [hunk 8]
```diff
@@ -10,6 +10,26 @@
...
+        raise ValueError(f"Provided method '{method}' is not supported.")
+
+    sparse_features = False
+    if isinstance(features, scipy.sparse.spmatrix):
+        features = features.toarray()
+        sparse_features = True
+
+    sentence_features = np.mean(features, axis=0)
+    if sparse_features:
+        return scipy.sparse.coo_matrix(sentence_features)
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/featurzier.py]: don't we use `csr_matrix` everywhere

File: rasa/nlu/featurizers/featurzier.py [hunk 9]
```diff
@@ -10,6 +10,26 @@
...
+) -> Optional[Union[np.ndarray, scipy.sparse.spmatrix]]:
+    if features is None:
+        return None
+
+    if method != "mean":
+        raise ValueError(f"Provided method '{method}' is not supported.")
+
+    sparse_features = False
+    if isinstance(features, scipy.sparse.spmatrix):
+        features = features.toarray()
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/featurzier.py]: let's not do that

File: rasa/nlu/featurizers/featurzier.py [hunk 10]
```diff
@@ -12,15 +12,20 @@
...
         return None
 
+    sparse_features = False
     if isinstance(features, scipy.sparse.spmatrix):
         features = features.toarray()
+        sparse_features = True
 
     if method == "mean":
-        return np.mean(features, axis=0)
+        sentence_features = np.mean(features, axis=0)
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/featurizers/featurzier.py]: sorry, this wrong sum method

[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/featurizers/featurzier.py]: it looks like `car_matrix` supports `+` operation, so python `sum([..])` should work

[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/featurizers/featurzier.py]: why do you want to use sum? the sentence vector of spacy is calculated by taking the mean of the token vectors.

File: rasa/nlu/featurizers/featurzier.py [hunk 11]
```diff
@@ -11,16 +11,13 @@
...
 ) -> Optional[Union[np.ndarray, scipy.sparse.spmatrix]]:
     if features is None:
         return None
 
-    if method != "mean":
-        raise ValueError(f"Provided method '{method}' is not supported.")
-
     if isinstance(features, scipy.sparse.spmatrix):
-        return scipy.sparse.csr_matrix(features.mean(axis=0))
+        return scipy.sparse.csr_matrix(features.sum(axis=0))
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/featurzier.py]: do we need to wrap it in `scipy.sparse.csr_matrix`?

Comment (Ghostvv) [reply rasa/nlu/featurizers/featurzier.py]: yes, we need, it returns `numpy`

**Pull request merged**